### ISSUE TITLE
Inquiry: Support for `self.training` in Burn Modules or Alternative Solutions

### ISSUE THREAD

**Issue by yiwufen** [tracel-ai/burn#2652]
### Inquiry: Support for `self.training` in Burn Modules or Alternative Solutions

**Description**

Hello Burn community,

I am currently developing a model using the [Burn](https://burn.rs/) framework and have encountered a challenge related to managing training and evaluation modes within the `forward` method. In PyTorch, we utilize the `self.training` attribute to differentiate between training and evaluation phases, allowing for conditional logic based on the current mode. I am seeking guidance on how to achieve similar functionality in Burn or if there are recommended alternative approaches.

**PyTorch Implementation Example**

In PyTorch, the `self.training` attribute is commonly used within the `forward` method to switch behaviors between training and evaluation modes. Here is an example of how it's implemented:

```python
import torch.nn as nn

class CVRPModel(nn.Module):
    def __init__(self):
        super(CVRPModel, self).__init__()
        self.decoder = Decoder()

    def forward(self, state):
        probs = self.decoder(state)
        if self.training:
            # Training mode logic, e.g., using sampling
            selected = torch.multinomial(probs, 1)
        else:
            # Evaluation mode logic, e.g., selecting the highest probability
            selected = torch.argmax(probs, dim=1)
        return selected
```

In this example:
- `self.training` determines whether the model is in training mode.
- `torch.multinomial` is used during training for stochastic selection.
- `torch.argmax` is used during evaluation for deterministic selection.

**Questions**

1. **Does Burn Support `self.training`-Like Functionality?**
   - Is there an equivalent attribute or method in Burn that allows modules to determine if they are in training or evaluation mode within the 

forward

 method?

2. **Implementing Conditional Logic Based on Mode**
   - If Burn does not have a direct equivalent to `self.training`, what is the recommended approach to implement conditional forward logic based on the training or evaluation phase?

3. **Example Conversion**
   - How can the provided PyTorch 

forward

 method be translated to Burn to achieve similar functionality? Below is an attempt to convert the PyTorch code to Burn, but I'm unsure if this is the correct approach:

```rust
impl<B: Backend> CVRPModel<B> {
    pub fn forward(&self, state: StepState<B>) -> Tensor<B, 2> {
        let probs = self.decoder.forward(state);
        let selected = if self.training {
            // Training mode logic, e.g., using sampling
            probs.multinomial(1)
        } else {
            // Evaluation mode logic, e.g., selecting the highest probability
            probs.argmax(&[2], true)
        };
        selected
    }
}
```

**Additional Information**

- In PyTorch, `self.training` is controlled by calling `model.train()` for training mode and `model.eval()` for evaluation mode.
- I would like to know if Burn provides similar mechanisms to switch between training and evaluation modes and how to access the current mode within module implementations.

**Expected Outcome**

Guidance on how to:
- Utilize training and evaluation modes within Burn modules.
- Implement conditional logic in the 

forward

 method based on the current mode.
- Properly translate PyTorch training-mode-dependent logic to Burn.

**Thank You**

Thank you for your assistance and for developing such an impressive framework. I look forward to your guidance on this matter!

```

**Comment by laggui** [comment]
Just check if autodiff is enabled on the backend! This will only be true for the autodiff backend (used for training, see the [difference with pytorch](https://burn.dev/burn-book/building-blocks/autodiff.html#difference-with-pytorch)).

To answer your question specifically, take a look at how this is done for batchnorm, which has a different behavior for training vs inference:

https://github.com/tracel-ai/burn/blob/2338912dafb7f9f18c8d14b851f7c7fd6ec542c7/crates/burn-core/src/nn/norm/batch.rs#L86-L103

So in your case, it would be something like this:

```rust
impl<B: Backend> CVRPModel<B> {
    pub fn forward(&self, state: StepState<B>) -> Tensor<B, 2> {
        let probs = self.decoder.forward(state);
        match B::ad_enabled() {
            true =>  probs.multinomial(1),
            false => probs.argmax(&[2], true),
    }
}
```
